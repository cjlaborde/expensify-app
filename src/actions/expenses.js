 import uuid from 'uuid'

// ADD_EXPENSE
// It's set up to take object with these attributes from AddExpensePage->ExpenseForm
export const addExpense = (
    {
        // Default Values
        description = '',
        note = '',
        amount = 0, 
        createdAt = 0 
    } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        //shorthand Es6 Method
        description,
        note,
        amount,
        createdAt
    }
})

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})