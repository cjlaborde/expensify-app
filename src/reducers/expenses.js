// 1) Expenses Reducer

const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
        return [
            ...state,
            action.expense
        ]
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    // if matched we want to return a new object
                    return {
                        // object spread operator
                        // grab all properties from existing one 
                        ...expense,
                        // overwrite the ones that were pass in. In this case store.dispatch(editExpense(expenseTwo.expense.id, { amount: 500 }))
                        // create new array
                        ...action.updates
                    }

                } else {
                    return expense // means we do nothing since there was no match
                }
            })
        default: 
        return state
    }
}

export default expensesReducer