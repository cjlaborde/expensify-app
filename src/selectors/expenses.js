import moment from 'moment'

// Get visible expenses

// You have to detructure filters  as done below
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        // https://momentjs.com/docs/#/query/is-same-or-before/
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        
        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {  //https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}