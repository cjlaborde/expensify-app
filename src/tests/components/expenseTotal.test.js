import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'
import ExpenseSummary from '../../components/ExpensesSummary'


test('should render ExpenseList with expenses', () => {
    // requires props so you need to add it
    const wrapper = shallow(<ExpenseSummary expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})


test('should return 0 if no expenses', () => 
{
    const wrapper = shallow(<ExpenseSummary expenses={expenses} />)
    expect(wrapper)
    
})

test('should correctly add up a single expense', () => {

})

test('should correctly add up multiple expense', () => {

})

/*

let addExpense, history, wrapper

beforeEach(() => {
// we use same since we still rendering
    addExpense = jest.fn()
    history = { push: jest.fn() }
    // use spies defined above
    wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history} />)
})

test('should render AddExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot()
})
*/