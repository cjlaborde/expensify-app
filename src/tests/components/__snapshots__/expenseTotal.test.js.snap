// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
